pex/pip.py

```diff
@@ -610,18 +633,179 @@ def create(
+                                    def versions_to_string(versions):
+                                        return [".".join(map(str, version)) for version in versions]
+
+                                    python_versions_strings = versions_to_string(
+                                        python_versions
+                                    ) or skip
+
+                                    python_full_versions_strings = versions_to_string(
+                                        python_full_versions
+                                    ) or skip
+
+                                    def _get_env(environment, name):
+                                        if name == "extra":
+                                            return original_get_env(environment, name)
+                                        if name == "python_version":
+                                            return python_versions_strings
+                                        if name == "python_full_version":
+                                            return python_full_versions_strings
+                                        return skip
+
+                                    def _eval_op(lhs, op, rhs):
+                                        if lhs is skip or rhs is skip:
+                                            return True
+                                        return any(
+                                            original_eval_op(left, op, right)
+                                            for left in (lhs if isinstance(lhs, list) else [lhs])
+                                            for right in (rhs if isinstance(rhs, list) else [rhs])
+                                        )
+
+                                    markers._get_env = _get_env
+                                    markers._eval_op = _eval_op
+                                
+                                patch_marker_evaluate()
+                                del patch_marker_evaluate
+
+                                # 2.) Universal wheel tag applicability.
+                                #
+                                # Allows all wheel URLs to be checked even when the wheel does not
+                                # match system tags.
+                                def patch_wheel_model():
+                                    from pip._internal.models.wheel import Wheel
+
+                                    Wheel.support_index_min = lambda *args, **kwargs: 0
+
+                                    if python_versions:
+                                        import re
+
+                                        def supported(self, *_args, **_kwargs):
+                                            if not hasattr(self, "_versions"):
+                                                versions = set()
+                                                for pyversion in self.pyversions:
+                                                    if pyversion[:2] in ("cp", "pp", "py"):
```

Comment:
https://github.com/pantsbuild/pex/runs/4933490418?check_suite_focus=true#step:7:935 exposed a bug here. This only considers `<python>` from the `<python>-<abi>-<platfrom>` tag triple, but when `<abi>` is abi3, `<python>` evaluation should only consider the major version of `<python>` - namely `3`. So, right now, since this lock is for `==3.9.*` the compatible `cryptography-35.0.0-cp36-abi3-macosx_10_10_x86_64.whl` wheel if filtered out and an attempt to build cryptography with the sdist proceeds and fails.

Reply:
> evaluation should only consider the major version of <python> - namely 3.

It also needs to consider the <python> prefix if `--interpreter-constraint`, right? `cp36-abi3` is compatible with 3.6+ but not 3.5.

Reply:
Can you cite your source for that? Pip (packaging 20.8 really), disagrees (Here my `pip` tox env is Python 3.10.):
```
$ tox -epip -- debug -v | grep abi3 | grep linux_x86_64
  cp310-abi3-linux_x86_64
  cp39-abi3-linux_x86_64
  cp38-abi3-linux_x86_64
  cp37-abi3-linux_x86_64
  cp36-abi3-linux_x86_64
  cp35-abi3-linux_x86_64
  cp34-abi3-linux_x86_64
  cp33-abi3-linux_x86_64
  cp32-abi3-linux_x86_64
```
 So packaging takes it back from present to the birth of abi3 in Python 3.2.

Reply:
Keep in mind, over inclusion is fine (you've already signed away your right to worry about un-vetted dists by choosing universal); so If there is a cp3{0,1} wheel out there that claims abi3; so be it. At resolve time, packaging will be used to evaluate tags and reject such a tag and so your resolve will either fail (again - you bough into this by choosing universal), or you'll proceed to get the sdist and build that (another thing you bought into).

Reply:
>> cp36-abi3 is compatible with 3.6+ but not 3.5.

> Can you cite your source for that?

PEP 652: https://www.python.org/dev/peps/pep-0652/#stable-abi

> The Stable ABI is not generally forward-compatible: an extension built and tested with CPython 3.10 will not generally be compatible with CPython 3.9.

> At resolve time, packaging will be used to evaluate tags and reject such a tag and so your resolve will either fail (again - you bough into this by choosing universal)

I wonder if it's possible to reject the lock, i.e. more eager than failing at resolution time. If you set `--interpreter-constraint="CPython==3.6.*"` and there is only a `cp38-abi3` wheel available, it would be neat if we could fail to lock.

Reply:
That's the wrong PEP / standard. The _Limited_ ABI is what abi3 is all about:
+ https://www.python.org/dev/peps/pep-0652/#stable-abi-vs-limited-api
+ https://www.python.org/dev/peps/pep-0384

The fact Pip / packaging emit my sample tag set above for a Python 3.10 interpreter implies they view things this way at least.

Reply:
> ... it would be neat if we could fail to lock.